README.md
# LuxPower Distribution Card

[![hacs_badge](https://img.shields.io/badge/HACS-Default-41BDF5.svg?style=flat-square)](https://github.com/hacs/integration)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/DanteWinters/lux-power-distribution-card?style=flat-square)
![Github stars](https://img.shields.io/github/stars/DanteWinters/lux-power-distribution-card?style=flat-square)
![Github issues](https://img.shields.io/github/issues/DanteWinters/lux-power-distribution-card?style=flat-square)

A simple power distribution card of an inverter and battery system, for [Home Assistant](https://home-assistant.io/). The card is modelled after LuxpowerTek's app and website.

<img src="https://raw.githubusercontent.com/DanteWinters/lux-power-distribution-card/main/docs/images/full-card-allocated-power.png" width="450" />

## Installation

### HACS

This card has been added to the list of default HACS frontend elements. Search for the name on HACS, and download from there to install.

## Adding the card to the dashboard

### Configuration

The following is a list of configs for the card:

**NOTE:** Please refer to the example config below for clarification on how the entities should be added.

#### Required configurations

| Name | Type | Description |
|---|:---:|---|
| inverter_count | number | The number of inverters used in the system. Must be a positive integer that is 1 or higher. |
| battery_soc | entities | Battery state of charge. |
| battery_flow | entities | Power flowing from and to the battery. Negative flow is discharge, and positive flow is charge. |
| home_consumption | entities | Output power of the inverter to your home. |
| grid_flow | entities | Power flowing to and from grid. Negative flow is import from grid, and positive flow is export to grid. |

#### Optional configurations

| Name | Type | Description |
|---|---|---|
| battery_voltage | entities | Battery's voltage. |
| pv_power | entities | Solar power. |
| backup_power | entities | This is off-grid power. In the case of the LuxpowerTek inverter, this is used when grid is not available. |
| grid_voltage | entities | Grid's voltage. |
| energy_allocations | entities | This is not a single entity, but a list of entities. Explanation below. |
| update_time | entities | An entity for the last time the values were updated. |
| status_codes | entities | An entity that has the inverter status. This is directly used with the LuxpowerTek integration (See below).  |
| lux_dongle | list of strings | This is the LuxPower inverter's dongle number. It will later on be used to call the refresh service. (This requires the LuxPowerTek integration that supports this.) This or the inverter alias list are required if there are more than 1 inverter. |
| inverter_alias | list of strings | This is used when there is more than 1 inverter. This will be the names used in the dropdown list. This or the lux dongle list is required. |
| refresh_button | string | The location of the refresh button. Can be 'left', 'right' or 'both'. See below for more information. **NOTE:** the refresh button will only show if the *lux_dongle* is added. |
| title | string | A title for the card. |

#### Sub-configs that are not a list of entities or values

| Parent | Name | Type | Description |
|---|---|---|---|
| grid_indicator | hue | bool | If this is set to true and the grid voltage drops to 0, the grid image will become dimmer. (Requires a grid voltage entity.) |
| grid_indicator | dot | bool | If this is set to true and the grid voltage drops to 0, a red indicator will be added next to the grid voltage text. (Requires a grid voltage entity.) |
| update_time | show_last_update | bool | If the update time entity has a timestamp attribute, it can be used to show how long since the last update. |
| status_codes | no_grid_is_warning | bool | Some status codes (64, 136 and 192) are shown when grid is not available. If this value is true, these codes will show up as a warning on the status. If the value is false, these values will show up as normal. |
| parallel | average_voltage | bool | When using multiple inverters, there is a default created item on the list of inverters called "Parallel" that averages all the values from the different inverters. If *average_voltage* is true, the battery and grid voltages will be averaged and shown on the Parallel setting. Otherwise it will not show the voltages there. |
| parallel | parallel_first | bool | When using multiple inverters, there is a default created item on the list of inverters called "Parallel" that averages all the values from the different inverters. If *parallel_first* is true, the "Parallel" option will be shown first of the list, otherwise it will be last. |

### Example Configuration

If you have the LuxpowerTek integration, you can use the following code directly (except for the energy_allocations, and change the dongle number):

 ```yaml
type: custom:lux-power-distribution-card
inverter_count: 1
parallel:
  average_voltage: true
  parallel_first: true
battery_soc:
  entities:
    - sensor.lux_battery
battery_flow:
  entities:
    - sensor.lux_battery_flow_live
home_consumption:
  entities:
    - sensor.lux_home_consumption_live
grid_flow:
  entities:
    - sensor.lux_grid_flow_live
lux_dongle:
  - BAxxxxxxxx
inverter_alias:
  - Inverter 1
battery_voltage:
  entities:
    - sensor.lux_battery_voltage_live
pv_power:
  entities:
    - sensor.lux_solar_output_live
backup_power:
  entities:
    - sensor.lux_power_to_eps_live
grid_voltage:
  entities:
    - sensor.lux_grid_voltage_live
update_time:
  entities:
    - sensor.lux_data_received_time
  show_last_update: true
grid_indicator:
  hue: true
status_codes:
  entities:
    - sensor.lux_status
  no_grid_is_warning: true
refresh_button: right
energy_allocations:
  entities:
    - sensor.power_plug_1
    - sensor.power_plug_2
    - sensor.power_plug_3
    - sensor.power_plug_4
```

## LuxpowerTek integration

The LuxpowerTek integration is hosted in a private repository by [Guy Wells](https://github.com/guybw).

## Refresh and the Dongle serial number

This refresh only works for the LuxPowerTek integration referenced above. The service name and function call format are hard-coded.

The location of the refresh button can be set with the *refresh_button_location* config value. There are 4 accepted values for this config:

- left (Displayed on the left hand side, below the battery text.)
- right (Displayed on the right hand side, above the grid image.)
- both (Displayed on both sides, as described on the above two points.)
- none (Removes the refresh button.)

## Energy Allocations Entities

The *energy_allocations* entities can be any entity that measures power. It will sum the values together and display on the card. The idea is to use this to track how much of the home's power usage is known.

## Grid indicators

Below are 2 pictures of the grid image. The first is the grid in a normal state, and the second is the grid image with both indicators active.

| Normal Grid | No Grid Input |
|---|---|
| <img src="https://raw.githubusercontent.com/DanteWinters/lux-power-distribution-card/main/docs/images/grid-normal.png" /> | <img src="https://raw.githubusercontent.com/DanteWinters/lux-power-distribution-card/main/docs/images/grid-no-ac.png" /> |

## Gallery

| The card with only required entities | The card with all required and optional entities | The card using all the LuxPower integration options and entities |
|---|---|---|
| <img src="https://raw.githubusercontent.com/DanteWinters/lux-power-distribution-card/main/docs/images/base-card.png" /> | <img src="https://raw.githubusercontent.com/DanteWinters/lux-power-distribution-card/main/docs/images/base-card-with-extras.png" /> | <img src="https://raw.githubusercontent.com/DanteWinters/lux-power-distribution-card/main/docs/images/full-card.png" /> |

## Interactive Card

The four entity images on the card can be clicked to display the history of the connected entity.

- Solar Image: Solar power entity's history.
- Battery Image: Battery SOC entity's history.
- Home Image: Home consumption entity's history.
- Grid Image: Grid flow entity's history.

## Parallel inverters

With v1.0.0, support for parallel inverters have arrived! In order to use parallel inverters, simply indicate the number of inverters you are using in the config, and add the additional inverter's entities under their corresponding headers. Take note of the *inverter_alias* and *lux_dongle* config values when using parallel inverters.

The status bar for the parallel inverters works as follows:

- If both inverters have normal status, it will display a normal status.
- If only one inverter has an non-normal status, the inverter alias will be displayed along with the non-normal status.
- If all the inverters have the same error (i.e. 'no-grid'), it will display this error on the parallel page.
- If there are multiple different error, the status will display as 'multiple errors' and you will need to go to the specific inverter to see the error.

## Known issues

### Card not loading issue

With this card, there has been multiple instances of the card not loading. From my experience, the best way to fix this is to clear the cache and it should load. I can give instructions for both Android mobile devices and the browser.

#### Android

1. Find *Home Assistant* in the list of apps in settings.
2. This step may differ depending on the Android device. Find anything that indicates data used or storage.
3. When there, find the option to clear all the data (cache and storage). Clearing this will log you out of the app and you'll need to log in again.
4. Card should then show up. If it doesn't, please log a bug.

#### Web browser

1. With the page open, open the developer console on the browser. Usually it's *F12*.
2. Click on the refresh button.
3. Rick-click on the refresh button to open a menu.
4. Choose the option *Empty Cache and Hard Reload*, or the option closest to this description.

## Developer's note

Although the card is functional and even has a few nice features, the development of it was done with a lot of inexperience. From my side, I do not have JavaScript or HTML experience other than this card. For this reason, there may be many ways I implemented things that aren't optimal or safe. If you are knowledgeable in and willing to look through the code, and advice and help will be much appreciated.

In addition, I currently only have 1 inverter. So the tests for the parallel inverters were done purely in a testing environment and may have some bugs.


docs/changelog.md
## Changelog

### v0.1.0

- Initial Release

### v0.2.0

- Reworked the card with better practices.

### v0.2.2

Enhancements

- Implemented HACS validations actions.
- Updated README for HACS requirements.
- Moved README to root folder.

Bugfixes

- Fixed arrow-animation issus
- Cropped arrow image, final result is that the arrows are larger and closer together in the final card.

### v0.3.0

Enhancements

- The refresh button can be shows by adding the correct config, but unfortunately the service isn't called when pressing the button. When this issue is solved, the rest of the interactions should be simple
- Images stored as base64. Issue can be closed when v03.0 is made the latest release.
- Keeping the README up to date is a continuous process but it is sufficient for use and HACS.
- PR for adding to HACS is pending, but the card can be added as a custom repository.
- There are now 3 ways of showing grid status. The LuxPower integration will still require fine tuning.
- Added label to show when last the values were updated, and also how long ago that was (if the entity has a timestamp attribute)

Bugfixes

- Improved formatting and scaling but text is not at a point where I am satisfied with.

### v0.4.0

Features

- Added the functionality to see the entity history and the refresh button works.

### v0.4.2

Bugfix

- Changed styles to accommodate safari browsers.

### v1.0.0

Features

- Parallel inverters
  - v1.0.0 implements parallel inverters. Adding a second inverter will allow you to choose which inverter's info you want to see. Blending the info is the next step
- Card now uses status codes directly from the integration and gives a short description based on that.

Breaking changes

- v1.0.0 implements a new config format. This will break existing cards until the new config is implemented. Please refer to the README file for information.

### v1.1.0

Features

- Added mixing between parallel inverters
- Refresh button on Parallel page will. refresh both inverters

### v1.1.1

Bugfix

- For parallel page, all values are added except for battery SOC and voltages.

### v1.1.2

Bugfixes

- Round values to max of 2 decimal places.
- Reworked code to fix issue with battery arrows not showing.
- Reverted LitElement to HTMLElement.

### v1.2.0

Features

- Updated the status indicator to allow for parallel inverters.


config-entity-functions.js
import * as constants from "./constants.js";

const required_entries = ["inverter_count", "battery_soc", "battery_flow", "home_consumption", "grid_flow"];

export function buildConfig(config) {
  function deepcopy(value) {
    if (!(!!value && typeof value == "object")) {
      return value;
    }
    if (Object.prototype.toString.call(value) == "[object Date]") {
      return new Date(value.getTime());
    }
    if (Array.isArray(value)) {
      return value.map(deepcopy);
    }
    const result = {};
    Object.keys(value).forEach(function (key) {
      result[key] = deepcopy(value[key]);
    });
    return result;
  }

  config = deepcopy(config);
  const new_config = deepcopy(constants.base_config);

  new_config.title = config.title;

  // Check inverter count
  let inverter_count = parseInt(config.inverter_count);
  if (isNaN(inverter_count) || inverter_count <= 0) {
    throw new Error(
      "You don't have the 'inverter_count' config. This likely means that you haven't updated to the new config used from v0.5.0 and on. Please take a look at the project's README on GitHub for more info."
    );
  }

  // Assign inverter count
  new_config.inverter_count = inverter_count;

  let config_entities = [
    "battery_soc",
    "battery_flow",
    "home_consumption",
    "grid_flow",
    "pv_power",
    "battery_voltage",
    "backup_power",
    "grid_voltage",
    "update_time",
    "status_codes",
  ];
  for (let i = 0; i < config_entities.length; i++) {
    importConfigEntities(config, new_config, inverter_count, config_entities[i]);
  }

  let config_values = ["lux_dongle", "inverter_alias"];
  for (let i = 0; i < config_values.length; i++) {
    importConfigValues(config, new_config, inverter_count, config_values[i]);
  }

  // Check last time on update time
  if (new_config.update_time.is_used) {
    if (config.update_time) {
      new_config.update_time.show_last_update = config.update_time.show_last_update;
    }
  }

  // Check grid indicator
  if (config.grid_indicator) {
    if (config.grid_indicator.hue) {
      new_config.grid_indicator.hue = true;
    }
    if (config.grid_indicator.dot) {
      new_config.grid_indicator.dot = true;
    }
  }

  // Check refresh button
  if (new_config.lux_dongle.is_used) {
    if (config.refresh_button) {
      switch (config.refresh_button) {
        case "left":
          new_config.refresh_button.left = true;
          break;
        case "right":
          new_config.refresh_button.right = true;
          break;
        case "both":
          new_config.refresh_button.left = true;
          new_config.refresh_button.right = true;
          break;
        default:
          break;
      }
    }
  }

  // Check on status with no grid
  if (new_config.status_codes.is_used) {
    if (config.status_codes && config.status_codes.no_grid_is_warning) {
      new_config.status_codes.no_grid_is_warning = config.status_codes.no_grid_is_warning;
    }
  }

  // Check allocated energy
  if (config.energy_allocations && config.energy_allocations.entities) {
    if (config.energy_allocations.entities.length > 0) {
      new_config.energy_allocations.is_used = true;
      new_config.energy_allocations.entities = config.energy_allocations.entities;
    }
  }

  // Check parallel settings
  if (config.parallel) {
    if (config.parallel.average_voltage) {
      new_config.parallel.average_voltage = true;
    }
    if (config.parallel.parallel_first) {
      new_config.parallel.parallel_first = true;
    } else {
      new_config.parallel.parallel_first = false;
    }
  }

  validateConfig(new_config);

  return new_config;
}

function validateConfig(config) {
  if (!config.battery_soc.is_used) {
    throw new Error("You need to define entities for the battery SOC (Equal to the inverter count).");
  }

  if (!config.battery_flow.is_used) {
    throw new Error("You need to define entities for the battery flow (Equal to the inverter count).");
  }

  if (!config.grid_flow.is_used) {
    throw new Error("You need to define entities for the grid flow (Equal to the inverter count).");
  }

  if (!config.home_consumption.is_used) {
    throw new Error("You need to define entities for the home consumption (Equal to the inverter count).");
  }

  if (config.inverter_count > 1) {
    if (!config.lux_dongle.is_used && !config.inverter_alias.is_used) {
      throw new Error("You need to define entities enough dongles or aliases for each inverter.");
    }
    if (config.lux_dongle.is_used && !config.inverter_alias.is_used) {
      config.inverter_alias.is_used = true;
      config.inverter_alias.values = config.lux_dongle.values;
    }
  }
}

function importConfigEntities(config, new_config, inverter_count, object_name) {
  if (config[object_name]) {
    if (config[object_name].entities) {
      if (config[object_name].entities.length == inverter_count) {
        new_config[object_name].is_used = true;
        new_config[object_name].entities = config[object_name].entities;
      }
    }
  }
}

function importConfigValues(config, new_config, inverter_count, object_name) {
  if (config[object_name]) {
    if (config[object_name].length == inverter_count) {
      new_config[object_name].is_used = true;
      new_config[object_name].values = config[object_name];
    }
  }
}

export function getEntitiesState(config, hass, config_entity, index) {
  const entity_name = config[config_entity].entities[index];
  try {
    const entity = hass.states[entity_name];
    if (entity.state) {
      if (entity.state === "unavailable" || entity.state === "unknown") {
        return "-";
      } else {
        return entity.state;
      }
    }
    return "-";
  } catch (error) {
    handleEntityError(entity_name, config_entity);
  }
}

export function getEntitiesNumState(config, hass, config_entity, index, is_int = true, is_avg = false) {
  let value = 0;
  if (index == -1) {
    for (let i = 0; i < config.inverter_count; i++) {
      value += parseFloat(getEntitiesState(config, hass, config_entity, i));
    }
    if (is_avg) {
      value = value / config.inverter_count;
    }
  } else {
    value = parseFloat(getEntitiesState(config, hass, config_entity, index));
  }
  if (is_int) {
    return parseInt(value);
  }
  return Math.round(value * 100) / 100;
}

export function getEntitiesAttribute(config, hass, config_entity, attribute_name, index) {
  const entity_name = config[config_entity].entities[index];
  try {
    const entity = hass.states[entity_name];
    if (entity.attributes && entity.attributes[attribute_name]) {
      return entity.attributes[attribute_name];
    } else {
      return "-";
    }
  } catch (error) {
    handleEntityError(entity_name, config_entity);
  }
}

export function getEntitiesUnit(config, hass, config_entity, index) {
  const entity_name = config[config_entity].entities[index];
  try {
    const entity = hass.states[config[config_entity].entities[index]];
    if (entity.state) {
      if (isNaN(entity.state)) return "-";
      else return entity.attributes.unit_of_measurement ?? "";
    }
    return "";
  } catch (error) {
    handleEntityError(entity_name, config_entity);
  }
}

const handleEntityError = (config_entry, entity_name) => {
  if (required_entries.includes(config_entry)) {
    throw new Error(`Invalid entity: ${entity_name} for config_entry`);
  }
};

export function getStatusMessage(status_code, show_no_grid_as_warning) {
  var status_level = 0;
  var message = "Warning";

  switch (status_code) {
    case 0:
      message = `Standby`;
      status_level = 0;
      break;
    case 1:
      message = `Error`;
      status_level = 2;
      break;
    case 2:
      message = `Inverting (Programming)`;
      status_level = 1;
      break;
    case 4:
      message = `Normal`;
      status_level = 0;
      break;
    case 9:
      message = `Normal (Selling)`;
      status_level = 0;
      break;
    case 10:
    case 12:
    case 16:
    case 20:
    case 32:
    case 40:
      message = `Normal`;
      status_level = 0;
      break;
    case 17:
      message = `High temp`;
      status_level = 2;
      break;
    case 64:
    case 136:
    case 192:
      if (show_no_grid_as_warning) {
        message = `No grid`;
        status_level = 1;
      } else {
        message = `Normal`;
        status_level = 0;
      }
      break;
    default:
      message = `Unknown code ${status_code}`;
      status_level = 2;
      break;
  }

  var indicator = "";
  switch (status_level) {
    case 0:
      indicator = `🟢`;
      break;
    case 1:
      indicator = `🟠`;
      break;
    case 2:
    default:
      indicator = `🔴`;
      break;
  }

  return `${message} ${indicator}`;
}
